Tools/CISupport/ews-build/steps.py

```diff
@@ -6874,3 +6902,333 @@ def doStepIf(self, step):
+
+        rc = yield super().run()
+        if rc != SUCCESS:
+            return defer.returnValue(rc)
+
+        self.createResultMessage()
+
+        unexpected_results = self.getProperty('unexpected_failing_files', 0) or self.getProperty('unexpected_new_issues', 0) or self.getProperty('unexpected_passing_files', 0)
+        if self.expectations and unexpected_results:
+            # If there are unexpected results, rebuild without changes to verify causation
+            self.build.addStepsAfterCurrentStep([RevertAppliedChanges(exclude=['new*', 'scan-build-output*']), ScanBuildSmartPointer(baseline=True)])
+        elif unexpected_results:
+            # Only save the results if there are failures and it is not the first run
+            self.build.addStepsAfterCurrentStep([ArchiveStaticAnalyzerResults(), UploadStaticAnalyzerResults(), ExtractStaticAnalyzerTestResults(), DisplaySmartPointerResults()])
+        return defer.returnValue(rc)
+
+    def createResultMessage(self):
+        log_text = self.log_observer.getStdout()
+        match = re.search(r'^Total (new issues|unexpected issues): (\d+)', log_text, re.MULTILINE)
+        if match:
+            self.resultMsg += f"{match.group(2)} new issue{'s' if int(match.group(2)) > 1 else ''} "
+            self.setProperty('unexpected_new_issues', int(match.group(2)))
+        else:
+            self.setProperty('unexpected_new_issues', 0)
+
+        match = re.search(r'^Total (new files|unexpected failing files): (\d+)', log_text, re.MULTILINE)
+        if match:
+            self.resultMsg += f"{match.group(2)} failing file{'s' if int(match.group(2)) > 1 else ''} "
+            self.setProperty('unexpected_failing_files', int(match.group(2)))
+        else:
+            self.setProperty('unexpected_failing_files', 0)
+
+        match = re.search(r'^Total fixed issues: (\d+)', log_text, re.MULTILINE)
+        if match:
+            self.resultMsg += f"{match.group(1)} fixed issue{'s' if int(match.group(1)) > 1 else ''} "
+            self.setProperty('fixed_issues', int(match.group(1)))
+        else:
+            self.setProperty('fixed_issues', 0)
+
+        match = re.search(r'^Total (fixed files|unexpected passing files): (\d+)', log_text, re.MULTILINE)
+        if match:
+            self.resultMsg += f"{match.group(2)} fixed file{'s' if int(match.group(2)) > 1 else ''}"
+            self.setProperty('unexpected_passing_files', int(match.group(2)))
+        else:
+            self.setProperty('unexpected_passing_files', 0)
+
+    def getResultSummary(self):
+        status = ''
+        if self.resultMsg:
+            status = f'{self.resultMsg}'
+        elif self.results == SUCCESS:
+            status = 'Found no unexpected results'
+        if self.results != SUCCESS:
+            status += f' ({Results[self.results]})'
+
+        return {u'step': status}
+
+
+class DisplaySmartPointerResults(buildstep.BuildStep, AddToLogMixin):
```

Comment:
would be nice to have unit-test as well for this step (and making sure all the steps being added have unit-tests).

Reply:
Added!